Mathlib/Data/Num/Basic.lean

```diff
@@ -0,0 +1,705 @@
+  | pos a, pos b => pos (a + b)
+  | pos a, neg b => sub' a b
+  | neg a, pos b => sub' b a
+  | neg a, neg b => neg (a + b)
+#align znum.add ZNum.add
+
+instance : Add ZNum :=
+  ⟨ZNum.add⟩
+
+/-- Multiplication of `ZNum`s. -/
+protected def mul : ZNum → ZNum → ZNum
+  | 0, _ => 0
+  | _, 0 => 0
+  | pos a, pos b => pos (a * b)
+  | pos a, neg b => neg (a * b)
+  | neg a, pos b => neg (a * b)
+  | neg a, neg b => pos (a * b)
+#align znum.mul ZNum.mul
+
+instance : Mul ZNum :=
+  ⟨ZNum.mul⟩
+
+open Ordering
+
+/-- Ordering on `ZNum`s. -/
+def cmp : ZNum → ZNum → Ordering
+  | 0, 0 => eq
+  | pos a, pos b => PosNum.cmp a b
+  | neg a, neg b => PosNum.cmp b a
+  | pos _, _ => gt
+  | neg _, _ => lt
+  | _, pos _ => lt
+  | _, neg _ => gt
+#align znum.cmp ZNum.cmp
+
+instance : LT ZNum :=
+  ⟨fun a b => cmp a b = Ordering.lt⟩
+
+instance : LE ZNum :=
+  ⟨fun a b => ¬b < a⟩
+
+instance decidableLt : @DecidableRel ZNum (· < ·)
+  | a, b => by dsimp [LT.lt]; infer_instance
+#align znum.decidable_lt ZNum.decidableLt
+
+instance decidableLe : @DecidableRel ZNum (· ≤ ·)
+  | a, b => by dsimp [LE.le]; infer_instance
+#align znum.decidable_le ZNum.decidableLe
+
+end ZNum
+
+namespace PosNum
+
+/-- Auxiliary definition for `PosNum.divMod`. -/
+def divModAux (d : PosNum) (q r : Num) : Num × Num :=
+  match Num.ofZNum' (Num.sub' r (Num.pos d)) with
+  | some r' => (Num.bit1 q, r')
+  | none => (Num.bit0 q, r)
+#align pos_num.divmod_aux PosNum.divModAux
+
+/-- `divMod x y = (y / x, y % x)`. -/
+def divMod (d : PosNum) : PosNum → Num × Num
+  | bit0 n =>
+    let (q, r₁) := divMod d n
+    divModAux d q (Num.bit0 r₁)
+  | bit1 n =>
+    let (q, r₁) := divMod d n
+    divModAux d q (Num.bit1 r₁)
+  | 1 => divModAux d 0 1
+#align pos_num.divmod PosNum.divMod
+
+/-- Division of `PosNum`, -/
+def div' (n d : PosNum) : Num :=
+  (divMod d n).1
+#align pos_num.div' PosNum.div'
+
+/-- Modulus of `PosNum`s. -/
+def mod' (n d : PosNum) : Num :=
+  (divMod d n).2
+#align pos_num.mod' PosNum.mod'
+
+/-
+  private def sqrt_aux1 (b : pos_num) (r n : num) : num × num :=
+  match num.of_znum' (n.sub' (r + num.pos b)) with
+  | some n' := (r.div2 + num.pos b, n')
+  | none := (r.div2, n)
+  end
+
+  private def sqrt_aux : pos_num → num → num → num
+  | b@(bit0 b') r n := let (r', n') := sqrt_aux1 b r n in sqrt_aux b' r' n'
+  | b@(bit1 b') r n := let (r', n') := sqrt_aux1 b r n in sqrt_aux b' r' n'
+  | 1           r n := (sqrt_aux1 1 r n).1
+  -/
+/-
+
+def sqrt_aux : ℕ → ℕ → ℕ → ℕ
+| b r n := if b0 : b = 0 then r else
+  let b' := shiftr b 2 in
+  have b' < b, from sqrt_aux_dec b0,
+  match (n - (r + b : ℕ) : ℤ) with
+  | (n' : ℕ) := sqrt_aux b' (div2 r + b) n'
+  | _ := sqrt_aux b' (div2 r) n
+  end
+
+/-- `sqrt n` is the square root of a natural number `n`. If `n` is not a
+  perfect square, it returns the largest `k:ℕ` such that `k*k ≤ n`. -/
+def sqrt (n : ℕ) : ℕ :=
+match size n with
+| 0      := 0
+| succ s := sqrt_aux (shiftl 1 (bit0 (div2 s))) 0 n
+end
+-/
```

Comment:
Maybe just drop this? Or leave a comment to the effect that there is a commented out implementation of this in mathlib3??

Reply:
the second comment is definitely not needed, it is just a copy of the lean 3 nat.sqrt code, probably in preparation for a translation to num that never happened. The first one looks kind of plausible, does it actually work?

Reply:
The first comments work and I have it implemented in the current commit. @digama0 let me know if you want me to keep it, omit it, or comment it out.

Reply:
I'd say keep it. It's probably commented out because it's not so useful since it's just the aux function and the entry point is missing, but it's better to make sure it doesn't break in new versions of lean